File: tests/fixtures/ts_repo/src/shapes.ts
import { Logger } from './logger';

export interface Measurable {
  area(): number;
}

export interface Drawable extends Measurable {
  draw(): void;
}

export abstract class AbstractShape implements Drawable {
  protected readonly label: string;
  #hidden: number = 0;

  constructor(label: string, private readonly logger: Logger) {
    this.label = label;
  }

  abstract area(): number;

  draw(): void {
    this.logger.log(this.label);
  }
}

export class Rectangle extends AbstractShape implements Drawable {
  private box: BoundingBox;

  constructor(private width: number, private height: number, logger: Logger) {
    super('rectangle', logger);
    this.box = new BoundingBox(width, height);
  }

  area(): number {
    return this.width * this.height;
  }
}

export class BoundingBox {
  constructor(public width: number, public height: number) {}
}

export function totalArea(shapes: Measurable[]): number {
  return shapes.reduce((sum, shape) => sum + shape.area(), 0);
}

export type Maybe<T> = T | null;

File: tests/fixtures/java_repo/src/com/example/Shapes.java
package com.example;

import java.util.List;

/** Anything with an area. */
public interface Measurable {
    double area();
}

/** Anything that can be drawn. */
public interface Drawable extends Measurable {
    void draw();
}

/** Shared shape behaviour. */
public abstract class AbstractShape implements Drawable {
    protected final String name;
    private Logger logger;

    public AbstractShape(String name, Logger logger) {
        this.name = name;
        this.logger = logger;
    }

    public abstract double area();

    @Override
    public void draw() {
        logger.log(name);
    }
}

/** A rectangle. */
public class Rectangle extends AbstractShape implements Drawable {
    private double width;
    private double height;

    public Rectangle(double width, double height, Logger logger) {
        super("rectangle", logger);
        this.width = width;
        this.height = height;
    }

    @Override
    public double area() {
        return width * height;
    }
}
